pyartcd/pyartcd/pipelines/build_rhcos.py

```diff
@@ -0,0 +1,191 @@
+            # happen we will see a build start, assume we started it, and watch it to completion.
+            # this seems unlikely to cause any problems other than mild confusion.
+            self.start_build()
+            result["action"] = "build"
+            result["builds"] = self.wait_for_builds()
+
+        # final status in stdout
+        print(json.dumps(result))
+
+    def retrieve_auth_token(self) -> str:
+        """Retrieve the auth token from the Jenkins service account to use with Jenkins API"""
+        # use the first secret named after the jenkins service account (there can be several)
+        secret = next((s for s in oc.selector('secrets') if s.name().startswith('jenkins-token-')), None)
+        if secret is None:
+            raise Exception("Unable to find Jenkins service account token")
+
+        return base64.b64decode(secret.model.data.token).decode('utf-8')
+
+    @staticmethod
+    def build_parameters(build: Dict[str, List[Dict]]) -> Dict:
+        """Parse the build parameters from the build actions into a more usable dict"""
+        parameters = next((  # first action that has parameters
+            action["parameters"]
+            for action in build["actions"]
+            if "parameters" in action
+        ), [])
+        return {p["name"]: p["value"] for p in parameters}
+
+    @staticmethod
+    def build_url(job: str, number: int) -> str:
+        return f"{JENKINS_BASE_URL}/job/{job}/{number}/"
+
+    def query_existing_builds(self) -> List[Dict]:
+        """Check if there are any existing builds for the given version. Returns builds in progress."""
+        builds = []
+        for job in ("build", "build-arch", "release"):
+            builds.extend(
+                dict(**b, job=job, parameters=self.build_parameters(b), url=self.build_url(job, b["number"]))
+                for b in self.request_session.get(
+                    f"{JENKINS_BASE_URL}/job/{job}/api/json?tree=builds[number,description,result,actions[parameters[name,value]]]",
+                ).json()["builds"]
+                if b["result"] is None  # build is still running when it has no status
+            )
+
+        return [b for b in builds if b["parameters"].get("STREAM") == self.version]
+
+    def start_build(self):
+        """Start a new build for the given version"""
+        # determine parameters
+        params = dict(STREAM=self.version, EARLY_ARCH_JOBS="false")
+        if self.build_always:
+            params["FORCE"] = "true"
+
+        # start the build
+        self.request_session.post(
+            f"{JENKINS_BASE_URL}/job/build/buildWithParameters",
+            data=params,
+        )
+
+        # wait for a related build to begin
+        initial_builds = []
+        while not initial_builds:  # may take a few seconds for the build to start
+            time.sleep(1)
+            initial_builds = self.query_existing_builds()
```

Comment:
occurs to me this should have a limit. all kinds of ways this can go wrong and leave us hanging. if we're waiting more than 15 minutes for a build to start, something is probably wrong and we should try again later.